### ISSUE TITLE
Escalate era_boost past its all-width float32-cast ceiling, or decide medium wins on feasibility

### ISSUE THREAD

**Issue by vltnkiz** [vltnkiz/numerai#52]
Part of #46

## Question

[Tune SGDRegressor's alpha for zemir_01's all-width linear-stage escalation, then run the full fit](https://github.com/vltnkiz/numerai/issues/51) cleared the **linear** stage's `all`-width memory ceiling — a tuned `train_sgd` (`alpha=0.3, eta0=0.0003`) completed a full-scale `--model linear --feature-set all` fit, cached at `runs/harness/20260831T193849Z-linear-all-sgd-a0.3-e0.0003-drop141/`. The full **ensemble** fit then hit a second, independent ceiling in the `era_boost` stage, which #51 diagnosed but did not act on.

**The blowup**: `xgboost/data.py`'s `pandas_transform_data` converts a DataFrame column by column, holding every column's float32 copy in one list simultaneously. At 3,414 columns x 2,746,268 train rows that is **37.5 GiB of float32 on top of X's own int8**, projecting to a ~57.7 GiB peak against a 50 GiB ceiling (measured: 20.18 GiB after train load, climbing smoothly at ~2.5 GiB/s to a guard kill at 42.46 GiB inside `XGBRegressor.fit`). `train_xgboost` hits it twice — once in the initial `.fit(X, y)`, and again in `model.predict(X)`, which the era-boost loop calls on each of its 40 iterations. This is structurally the identical blowup #49 found in `train_ridge`, in a different stage.

Needs deciding:

- **Whether to escalate at all, or call `all` infeasible here.** `all` now needs a bespoke memory escalation at *every* stage — Ridge->SGD for linear (#50/#51), and now XGBoost->streaming for era_boost — each trading exactness for feasibility. There is a real argument that this is more machinery than a width decision justifies shipping into the live submission path, and that the answer to #48 is "medium, on feasibility grounds." Settle this before building.
- **If escalating: the streaming design.** XGBoost's first-class path is `QuantileDMatrix` built from a `DataIter` yielding row-batches with `release_data=True` (documented verbatim for this case: "Set it to True if the data transformation (converting data to np.float32 type) is memory intensive"). Each batch's cast is chunk-sized and freed; what stays resident is the quantized gradient index (~1 byte/entry at `max_bin=256`, ~9.4 GiB). Needs `xgb.train()` on the native API rather than `XGBRegressor.fit()` (which takes no DMatrix), with `xgb_model=booster` still carrying the era-boost loop's incremental continuation.
- **Chunked prediction** for the loop's `model.predict(X)`, mirroring what `_ScaledSGDPredictor` already does for the linear stage. Whether `inplace_predict` on int8 slices suffices, or an explicit per-chunk float32 cast is needed.
- **Batch size**, and whether the quantized index plus X's int8 actually leaves headroom for the loop's 40 iterations, not just the first fit.
- **Validation bar.** Map #46's shared-code safety net applies with full force: `train_xgboost` drives the live `run_pipeline.py` submission path, so any change to it must reproduce #42/#45's existing `medium`-width baseline numbers before it ships. Whether byte-for-byte reproduction is achievable (a streaming `QuantileDMatrix` may quantize differently from the single-batch one) or whether a same-ballpark bar is the honest standard, mirroring #50's choice for SGD.

If the escalation lands: run the full `--model ensemble --feature-set all --linear sgd --sgd-alpha 0.3 --sgd-eta0 0.0003` fit and report what it measures, unblocking [Decide zemir_01's width: all vs medium, once all fits in memory](https://github.com/vltnkiz/numerai/issues/48)'s actual all-vs-medium comparison. If it does not, or is judged not worth shipping, that resolves #48 as "medium reconfirmed as final, `all` infeasible at 64GiB."

**Use `~/rss_guard.py` for all memory-boundary work here**, not `ulimit -v` — #51 established the latter measures virtual arenas rather than real pressure, and the former is the RSS-based, child-invisible cap the #47/#49/#50 safety note was asking for.

**Comment by vltnkiz** [comment]
**Progress update — decisions settled, escalation implemented, full-scale fits running.**

## A finding that reframed the ticket before any decision was taken

#51's `all`-width linear cache (`runs/harness/20260831T193849Z-linear-all-sgd-a0.3-e0.0003-drop141/`) had **never been scored** — `scripts/score_harness.py` refuses a linear-only cache, because `scoring_sweep()` always asks for an `era_boost` column. Scored it directly at `p=0` (no neutralization means no projection, so this needed neither the feature load nor the per-era lstsq; throwaway `~/score_linear_only.py`, `scores.csv` deliberately not written):

| run | width / stage | mean_corr | sharpe | mean_mmc | payout |
|---|---|---|---|---|---|
| `20260831T193849Z` | **all, linear (SGD α=0.3)** | **0.023105** | **1.570** | **+0.001926** | **0.010977** |
| `20260830T195206Z` | medium, linear (OLS) | 0.015508 | 1.134 | −0.001285 | 0.000833 |
| `20260830T195013Z` | medium, linear (SGD) | 0.004168 | 0.439 | +0.001248 | 0.005450 |

Against the medium ensemble table at the same `p=0`: `linear+era_boost_p0` = 0.007459, `era_boost_p0` = 0.008455. **The `all`-width linear stage alone, un-neutralized, already out-payouts the entire medium ensemble at the same neutralization level**, and flips MMC from negative to positive. It does not beat medium's best row anywhere (`era_boost_p1`, 0.016411) — that row is neutralized at p=1, and this one isn't.

This is not itself the answer to #48, but it says `all` is worth the escalation rather than being a memory-driven curiosity.

## Design facts established by probe (xgboost 3.4.1, `~/probe_xgb.sh`)

| probe | result |
|---|---|
| `QuantileDMatrix.slice()` | **fails** — *"Slicing DMatrix is not supported for external memory"* |
| `inplace_predict` on int8 DataFrame / int8 numpy | **works**, bit-identical to float32 input |
| single-batch vs multi-batch `QuantileDMatrix` | **bit-identical** predictions |
| `xgb_model=` continuation on a `ref=`-linked `QuantileDMatrix` | works |
| `XGBRegressor.fit` vs native `xgb.train` + single-batch QDM | **bit-identical** |

The batching-invariance is structural, not luck: features are int8 with exact support `{0,1,2,3,4}` (#50) and `max_bin=256` is far past 5 distinct values, so the quantile sketch is exact however the rows are split.

## Decisions

1. **Escalate directly, no intermediate measurement.** The cheaper option — scoring the existing `all`-linear cache under subset neutralization before building — was considered and declined: `all` has to fit in XGBoost either way, so the measurement would not change what gets built.
2. **Per-stage width is out of scope.** "Linear at `all`, era_boost at `medium`" sidesteps this ceiling entirely and needs no escalation, but it is a different question from the one this map asks. Recorded on #46's Out of scope, not pursued here.
3. **`colsample_bytree = 0.00123` for the `all`-width fit.** #42's protocol holds *absolute* features-sampled-per-tree roughly constant across widths rather than the fraction (`small`: 0.1 × 42 ≈ 4.2; `medium`: 4.2/780 ≈ 0.005), so `all` gets 4.2/3414 ≈ 0.00123. Worth flagging loudly: **`config.py`'s default is `0.1`, which at `all` width samples 341 features per tree — 81× production's ~4.** A comparison whose arms differ in both feature count and per-tree sampling regime cannot attribute its result to width. No earlier run is contaminated: every `all`-width run so far was linear-only, where colsample is inert.
4. **Full move to native `xgb.train`**, not a partial patch. Only the initial `.fit(X, y)` and the loop's 40 × `.predict(X)` provably blow up, but `QuantileDMatrix` cannot be sliced, so the worst-era refit must build its own matrix regardless — a partial rewrite would leave two dialects in one function for no saving.
5. **Validation bar: rerun and compare, not byte-for-byte reproduction.** Both widths are refit under the *new* streaming path and compared against each other, rather than gating on reproducing #42/#45's existing `medium` numbers. This is a deliberate loosening of #46's shared-code safety net. It is apples-to-apples for the width decision, and the probe suggests exactness is likely free anyway — but **the consequence is real and is recorded here: if `medium` wins, #44 would cut a live model over to a `train_xgboost` never checked against today's production fit.** Whoever picks up #44 should re-read this paragraph.
6. **Keep `ref=dtrain`, batch 200k rows.** Dropping `ref=` would free ~9.4GB, but a worst-era slice that never observes some value in some column would quantize that column differently — a silent per-iteration drift across 40 iterations. The budget (~25.5GB projected against the ~50GB cap) affords correctness here.

## Implementation (uncommitted, working tree, same "leftover infra" pattern as #47/#50/#51)

`zemir/models.py`:
- `import xgboost as xgb` replaces `from xgboost import XGBRegressor`.
- **`_XGBBatchIter(xgb.DataIter)`** — yields row-batches as float32 with `release_data=True`, so the cast transient is batch-sized (~2.7GB) instead of whole-matrix-sized (~37.5GB). Its `rows` argument selects a subset *per batch*, so the loop's worst-era refit never materializes even one 4.7GB int8 half-copy of `X`.
- **`_ChunkedBoosterPredictor`** — the tree-stage counterpart to `_ScaledSGDPredictor`. Needed independently of training: `predict_each` over the ~4M-row validation set is **~54GB as a single float32 conversion, a larger blowup than training's**. Passes int8 straight to `inplace_predict`, no explicit cast.
- **`train_xgboost`** rewritten onto `xgb.train` + `QuantileDMatrix`, returning `_ChunkedBoosterPredictor`. `nthread`/`seed` are the native spellings of `n_jobs=-1`/`random_state`; new `batch_rows=200_000` parameter.

No other file changed — `config.py` calls `train_xgboost(**config.xgboost)` and `predict_each` only needs `.predict`, both unaffected. `LIVE` still ships `small`/OLS; the width knobs stay CLI-only until #44.

**Smoke passed**: `--smoke --model ensemble --feature-set medium --colsample-bytree 0.005` → 276,135 rows, 40 eras, both models.

## Running now

Serially (never concurrent — #42's infra note), each under `~/rss_guard.py --limit-gb 46` rather than `ulimit -v` (#51):

1. `--model ensemble --feature-set medium --colsample-bytree 0.005`
2. `--model ensemble --feature-set all --colsample-bytree 0.00123 --linear sgd --sgd-alpha 0.3 --sgd-eta0 0.0003 --drop-columns-file ~/all_width_drop_columns.json`

Logs: `~/run_fits.log`, `~/guard_ens_medium.log`, `~/guard_ens_all.log`. Ticket stays open until both complete and are scored.

**Comment by vltnkiz** [comment]
**Resolved: the escalation was built and both full-scale fits cleared the memory ceiling. Scoring the comparison is handed to a successor ticket.**

## Decisions (grilled this session)

1. **Escalate directly**, no intermediate measurement — `all` has to fit in XGBoost either way.
2. **Per-stage width (linear at `all`, era_boost at `medium`) ruled out of scope** for this map — recorded on #46's Out of scope, not pursued here.
3. **`colsample_bytree = 0.00123`** for the `all`-width fit, per #42's constant-features-per-tree protocol (4.2/3414). Flagged loudly: `config.py`'s default `0.1` would sample 341 features/tree at `all` width — 81x production's ~4 — and is not safe to use there.
4. **Full move to native `xgb.train`**, not a partial patch — `QuantileDMatrix.slice()` is unsupported, so a partial rewrite would've left two dialects in one function.
5. **Validation bar: rerun-and-compare, not byte-for-byte reproduction of the medium baseline.** Deliberate loosening of #46's shared-code safety net — flagged for whoever picks up #44: if `medium` wins, its `train_xgboost` was never checked against today's production fit.
6. **Keep `ref=dtrain`, batch 200k rows** — budget affords correctness over the ~9.4GiB it would free.

## Implementation (uncommitted, working tree — same pattern as #47/#50/#51)

`zemir/models.py`: `_XGBBatchIter(xgb.DataIter)` (streams row-batches, `release_data=True`), `_ChunkedBoosterPredictor` (chunked `inplace_predict`, the tree-stage counterpart to `_ScaledSGDPredictor`), `train_xgboost` rewritten onto `xgb.train`/`QuantileDMatrix`.

Two memory fixes found by the first failed `all`-width attempt (breached the 46GiB guard at 46.04GiB, 844s in):
- `zemir/pipeline.py`'s `fit_models` now releases pyarrow's pool **before each trainer**, not once after all of them — measured 10.3GiB of already-freed loader pages otherwise held resident through both trainers.
- `zemir/models.py`'s era-boost loop now drops the previous iteration's `QuantileDMatrix` before building the next, instead of implicitly holding both (~4.4GiB) across the rebind.

A third, found scoring the result: `zemir/data.py`'s `load_validation_features` now releases pyarrow's pool after its read — measured 29.67GiB resident against a 13.81GiB logical frame at `all` width, the same retention pattern in a code path nobody had touched yet. This one isn't inert like the training-side fixes: it changes what `score_configs` can run to completion, not what it computes.

None of the three change any fitted value or score — they free already-unreferenced memory. The `medium` ensemble cache was fit before the training-side fixes landed; both are provably numerically inert (freeing freed pages, dropping a dead reference), so it was not refit.

## Results

**Both full-scale ensemble fits completed under the ~50GiB cap** (guarded by `~/rss_guard.py`, serial per #42's infra note):

| fit | peak RSS | elapsed | exit |
|---|---|---|---|
| `medium` (`20260901T105520Z-ensemble-medium`) | 38.45 GiB | 22 min | 0 |
| `all` (`20260901T113555Z-ensemble-all-sgd-a0.3-e0.0003-drop141`) | 42.33 GiB | 108 min | 0 |

This is #52's central question, answered: **`all` width now runs the full ensemble end to end at 64GiB**, where the pre-escalation path projected a ~57.7GiB peak against the ~50GiB cap.

`medium` scored cleanly (`scripts/score_harness.py`, cache untouched by this ticket's fixes):

```
                        eras  mean_corr    sharpe  smart_sharpe  mmc_eras  mean_corr_window  mean_mmc  mmc_sharpe  max_feature_corr   payout
config
era_boost_p1             647   0.019283  1.310202      0.567089        89          0.011477  0.003468    0.325402          0.015094 0.016411
era_boost_p0.75          647   0.022572  1.463572      0.569070        89          0.012001  0.002554    0.237608          0.108538 0.014746
linear+era_boost_p1      647   0.015799  1.074956      0.490320        89          0.010189  0.003116    0.248834          0.019619 0.014653
...
linear_p0.5              647   0.015508  1.133628      0.457492        89          0.004965 -0.001285   -0.134101          0.188867 0.000833
```
(full table on the run's `scores.csv`; baseline `linear_p0.5` payout 0.000833)

**`all` is not yet scored.** The first attempt breached the memory guard in 50s — diagnosed as the `load_validation_features` bug above, not the fitting/prediction path this ticket set out to fix. The fix was verified in isolation (29.67GiB -> 14.42GiB), and a retry was ~30 minutes in with no ETA (per-era `lstsq` against ~3,555 neutralizer columns vs medium's 780 — a much larger sweep) when this session was paused. **Handed to the successor ticket** rather than held open for an unbounded wait.

## What the successor ticket needs to know

- Check `~/guard_score_all2.log` and `~/score_all_retry.log` on the WSL box first — the retry may have finished on its own; if so, its `scores.csv` is the answer, nothing to rerun.
- If not, rerun: `python ~/rss_guard.py --limit-gb 46 --log <name> -- python scripts/score_harness.py 20260901T113555Z-ensemble-all-sgd-a0.3-e0.0003-drop141`.
- **Keep the invoking `wsl.exe` process attached for the run's full duration** — #46's Notes already warn a background/disowned WSL invocation doesn't survive its parent process exiting; this run is not detached, but don't `&`-background it further.
- Worth a look, not yet investigated: scoring neutralizes against all 3,555 `all`-width feature names, while the fit itself used only 3,414 (141 dropped in #45 as near-zero-variance/near-duplicate). Very likely harmless — those columns carry ~no explanatory power in the OLS projection either way — but unverified.
- Once scored: report the `all` comparison table against `medium`'s above, and hand both to #48 — the actual all-vs-medium call is #48's job, not this one's.

**Comment by vltnkiz** [comment]
Escalation built and both full-scale fits cleared the memory ceiling — scoring handoff continues in the successor ticket.
